go/cli/internal/tui/chat.go

```diff
@@ -0,0 +1,442 @@
+			if sCmd != nil {
+				cmds = append(cmds, sCmd)
+			}
+			return m, tea.Batch(cmds...)
+		}
+	case tickMsg:
+		if m.working {
+			m.updateStatus()
+			return m, m.tick()
+		}
+		return m, nil
+	case tea.WindowSizeMsg:
+		// Reserve space for input and separator
+		inputHeight := 3
+		if !m.showInput {
+			inputHeight = 0
+		}
+		sepHeight := 2 // extra line for status
+		vpHeight := msg.Height - inputHeight - sepHeight
+		if vpHeight < 5 {
+			vpHeight = 5
+		}
+		m.vp.Width = msg.Width
+		m.vp.Height = vpHeight
+		m.input.SetWidth(msg.Width)
+		return m, nil
+	case tea.KeyMsg:
+		switch msg.String() {
+		case "ctrl+c", "esc":
+			if m.cancel != nil {
+				m.cancel()
+			}
+			return m, tea.Quit
+		case "enter":
+			if !m.showInput {
+				return m, nil
+			}
+			if m.streaming {
+				return m, nil
+			}
+			text := strings.TrimSpace(m.input.Value())
+			if text == "" {
+				return m, nil
+			}
+			m.appendUser(text)
+			m.input.Reset()
+			return m, m.submit(text)
+		}
+	case a2aEventMsg:
+		m.appendEvent(msg.Event)
+		return m, m.waitNext()
+	case streamDoneMsg:
+		m.streaming = false
+		m.working = false
+		m.updateStatus()
+		return m, nil
+	}
+
+	m.input, cmd = m.input.Update(msg)
+	if cmd != nil {
+		cmds = append(cmds, cmd)
+	}
+	return m, tea.Batch(cmds...)
+}
+
+func (m *chatModel) View() string {
+	width := m.vp.Width
+	if width <= 0 {
+		width = 80 // default width if not yet sized
+	}
+	status := m.statusText
+	if status == "" {
+		status = ""
+	}
+	if m.working {
+		status = fmt.Sprintf("%s %s", m.spin.View(), status)
+	}
+	if m.showInput {
+		return lipgloss.JoinVertical(lipgloss.Left,
+			m.vp.View(),
+			theme.SeparatorStyle().Render(strings.Repeat("─", max(10, width))),
+			theme.StatusStyle().Render(status),
+			m.input.View(),
+		)
+	}
+	return lipgloss.JoinVertical(lipgloss.Left,
+		m.vp.View(),
+		theme.SeparatorStyle().Render(strings.Repeat("─", max(10, width))),
+		theme.StatusStyle().Render(status),
+	)
+}
+
+func (m *chatModel) submit(text string) tea.Cmd {
+	m.streaming = true
+	m.working = true
+	m.workStart = time.Now()
+	m.updateStatus()
+	ctx, cancel := context.WithCancel(context.Background())
+	m.cancel = cancel
+
+	params := protocol.SendMessageParams{
+		Message: protocol.Message{
+			Role:      protocol.MessageRoleUser,
+			ContextID: &m.sessionID,
+			Parts:     []protocol.Part{protocol.NewTextPart(text)},
+		},
+	}
+
+	ch, err := m.send(ctx, params)
+	if err != nil {
+		m.appendError(err)
+		m.streaming = false
+		cancel()
+		return nil
+	}
+	m.streamCh = ch
+	return tea.Batch(m.waitNext(), m.tick())
+}
+
+func (m *chatModel) waitNext() tea.Cmd {
+	ch := m.streamCh
+	if ch == nil {
+		return nil
+	}
+	return func() tea.Msg {
+		ev, ok := <-ch
+		if !ok {
+			return streamDoneMsg{}
+		}
+		return a2aEventMsg{Event: ev}
+	}
+}
+
+func (m *chatModel) appendUser(text string) {
+	m.appendLine(theme.UserStyle().Render("You:") + " " + text)
+}
+
+func (m *chatModel) appendEvent(ev protocol.StreamingMessageEvent) {
```

Comment:
This function uses json utilities to parse the data, but most of this data is available properly via the go types, so please update to use those.